File: source/renderer/Camera.java
package renderer;

import imgui.ImGui;
import math.*;
import utility.Timer;
import window.Window;

public class Camera {
    public Float3 position = new Float3();
    public float aspect = 16.0f / 9.0f;
    public float fov = 75.0f;
    public float near = 0.01f;
    public float far = 500.0f;
    public float speed = 2.0f;
    public float sensitivity = 0.1f;

    private Float3 forward = Float3.getNegZ();
    private boolean firstClick = true;
    private boolean camMoving = false;

    public Camera() {}

    public void updateAspect(Int2 frameSize) {
        aspect = ((float) frameSize.x) / frameSize.y;
    }

    public void setForward(Float3 forward) {
        this.forward = forward.norm();
    }

    public Float3 getForward() {
        return new Float3(forward);
    }

    public Float3 getRight() {
        return forward.cross(Float3.getPosY());
    }

    public void moveForward(float deltaTime) {
        position = position.add(forward.mul(speed * deltaTime));
    }

    public void moveBack(float deltaTime) {
        position = position.sub(forward.mul(speed * deltaTime));
    }

    public void moveRight(float deltaTime) {
        position = position.add(getRight().mul(speed * deltaTime));
    }

    public void moveLeft(float deltaTime) {
        position = position.sub(getRight().mul(speed * deltaTime));
    }

    public void moveUp(float deltaTime) {
        position = position.add(Float3.getPosY().mul(speed * deltaTime));
    }

    public void moveDown(float deltaTime) {
        position = position.sub(Float3.getPosY().mul(speed * deltaTime));
    }

    public void rotate(Int2 mousePos, Int2 frameCenter, float verticalAngleLimit) {
	    final Int2 delta = mousePos.sub(frameCenter);
        final Float2 rotation = new Float2(delta).mul(sensitivity);

        Float3 forwardVert = forward.rotate(-rotation.y, getRight());
        if (Math.abs(forwardVert.angle(Float3.getPosY()) - 90.0f) <= verticalAngleLimit) {
            forward = forwardVert;
        }

        forward = forward.rotate(-rotation.x, Float3.getPosY());
    }

    public Mat4 viewMatrix() {
        return Mat4.lookAt(position, position.add(forward), Float3.getPosY());
    }

    public Mat4 projectionMatrix() {
        return Mat4.perspective(fov, aspect, near, far);
    }

    public Mat4 matrix() {
        return projectionMatrix().mul(viewMatrix());
    }

    public void setDefaultMovement(Window window, Timer timer) {
        window.keyboard.w.onHold.add(() -> moveForward(timer.getDeltaT()));
        window.keyboard.a.onHold.add(() -> moveLeft(timer.getDeltaT()));
        window.keyboard.s.onHold.add(() -> moveBack(timer.getDeltaT()));
        window.keyboard.d.onHold.add(() -> moveRight(timer.getDeltaT()));
        window.keyboard.e.onHold.add(() -> moveUp(timer.getDeltaT()));
        window.keyboard.q.onHold.add(() -> moveDown(timer.getDeltaT()));

        window.keyboard.shift.onPress.add(() -> speed = 5);
        window.keyboard.shift.onRelease.add(() -> speed = 2);

        window.mouse.mmb.onPress.add(() -> {
            window.mouse.setHidden(true);
            camMoving = true;
        });

        window.mouse.mmb.onHold.add(() -> {
            if (camMoving) {
			    final Int2 frameCenter = window.getSize().div(2);

                if (!firstClick) {
                    rotate(window.mouse.getPosition(), frameCenter, 85);
                }
                firstClick = false;

                window.mouse.setPosition(frameCenter);
            }
        });

        window.mouse.mmb.onRelease.add(() -> {
            window.mouse.setHidden(false);
            firstClick = true;
            camMoving = false;
        });
    }
}

File: klib/source/render/scene/camera.cpp
#include "render/scene/camera.h"


void kl::camera::update_aspect_ratio(const int2& size)
{
    aspect_ratio = (float) size.x / size.y;
}

void kl::camera::set_forward(const float3& dir)
{
    forward_ = math::normalize(dir);
}

void kl::camera::set_up(const float3& dir)
{
    up_ = math::normalize(dir);
}

kl::float3 kl::camera::get_forward() const
{
    return forward_;
}

kl::float3 kl::camera::get_right() const
{
    return math::cross(up_, forward_);
}

kl::float3 kl::camera::get_up() const
{
    return up_;
}

void kl::camera::move_forward(const float delta_time)
{
    position = position + forward_ * (speed * delta_time);
}

void kl::camera::move_back(const float delta_time)
{
    position = position - forward_ * (speed * delta_time);
}

void kl::camera::move_right(const float delta_time)
{
    position = position + get_right() * (speed * delta_time);
}

void kl::camera::move_left(const float delta_time)
{
    position = position - get_right() * (speed * delta_time);
}

void kl::camera::move_up(const float delta_time)
{
    position = position + up_ * (speed * delta_time);
}

void kl::camera::move_down(const float delta_time)
{
    position = position - up_ * (speed * delta_time);
}

void kl::camera::rotate(const float2& mouse_pos, const float2& frame_center, const float vertical_angle_limit)
{
    const float2 rotation = (mouse_pos - frame_center) * sensitivity;
    const float3 forward_vert = math::rotate(forward_, get_right(), rotation.y);

    if (abs(math::angle(forward_vert, up_) - 90.0f) <= vertical_angle_limit) {
        set_forward(forward_vert);
    }

    set_forward(math::rotate(forward_, up_, rotation.x));
}

kl::float4x4 kl::camera::view_matrix() const
{
    return float4x4::look_at(position, position + forward_, up_);
}

kl::float4x4 kl::camera::projection_matrix() const
{
    return float4x4::perspective(field_of_view, aspect_ratio, near_plane, far_plane);
}

kl::float4x4 kl::camera::matrix() const
{
    return projection_matrix() * view_matrix();
}
